pkg/kp/pcstore/consul_store.go

```diff
@@ -308,6 +309,147 @@ func (s *consulStore) WatchPodCluster(id fields.ID, quit <-chan struct{}) <-chan
+	defer func() {
+		for _, handler := range clusterUpdaters {
+			close(handler)
+		}
+	}()
+
+	// TODO: change to a reality query when status endpoint is set up.
+	var prevResults WatchedPodClusters
+	for {
+		select {
+		case curResults := <-watchedRes:
+			if curResults.Err != nil {
+				s.logger.WithError(curResults.Err).Errorln("Could not sync pod clusters")
+				continue
+			}
+			// zip up the previous and current results, act based on the difference.
+			zipped := s.zipResults(curResults, prevResults)
+			for id, change := range zipped {
+				updater, ok := clusterUpdaters[id]
+				if !ok {
+					clusterUpdaters[id] = make(chan podClusterChange)
+					go s.handlePCUpdates(syncer, clusterUpdaters[id])
+					updater = clusterUpdaters[id]
+				}
+				select {
+				case updater <- change:
+					if change.previous != nil && change.current == nil {
+						close(clusterUpdaters[id])
+						delete(clusterUpdaters, id)
+					}
+				case <-quit:
+					return nil
+				}
+			}
+			prevResults = curResults
+		case <-quit:
+			return nil
+		}
+	}
+
+	return nil
+}
+
+// zipResults takes two sets of watched pod clusters and joins them such that they
+// are paired together in a map of pc ID -> change objects. Each change will be sent
+// to the respective sync channels of each pod cluster later on.
+func (s *consulStore) zipResults(current, previous WatchedPodClusters) map[fields.ID]podClusterChange {
+	allPrevious := make(map[fields.ID]*fields.PodCluster)
+	for _, prev := range previous.Clusters {
+		allPrevious[prev.ID] = prev
+	}
+	ret := map[fields.ID]podClusterChange{}
+	for _, cur := range current.Clusters {
+		prev, ok := allPrevious[cur.ID]
+		ret[cur.ID] = podClusterChange{
+			previous: prev,
+			current:  cur,
+		}
+		if ok {
+			delete(allPrevious, cur.ID)
+		}
+	}
+	for _, prev := range allPrevious {
+		ret[prev.ID] = podClusterChange{
+			previous: prev,
+		}
+	}
+	return ret
+}
+
+// try forever to match the expectations as defined in the provided change channel.
+// If a change fails to take, this function will retry that change forever until
+// it works as expected or a newer change appears on the channel. This routine also
+// executes and monitors the label watch for the pod's label selector.
+// TODO: since most of this code is generic to pod cluster syncers, this will likely
+// be extracted to github.com/square/p2.
+func (s *consulStore) handlePCUpdates(concrete ConcreteSyncer, changes chan podClusterChange) {
+	var change podClusterChange
+	podWatch := make(chan []labels.Labeled)
+	podWatchQuit := make(chan struct{})
+	defer func() {
+		close(podWatchQuit)
+	}()
+
+	for {
+		var ok bool
+
+		select {
+		case labeledPods := <-podWatch:
+			s.logger.Debugf("Calling SyncCluster with %v / %v", change.current, labeledPods)
+			concrete.SyncCluster(change.current, labeledPods)
+		case change, ok = <-changes:
+			if !ok {
+				return // we're closed for business
+			}
+
+			if change.current == nil && change.previous != nil {
```

Comment:
might be nice to define functions like `change.IsDeletion()` or maybe `change.Type()` that you can `switch` on